Make vttestserver compatible with persistent data directories

hkdsun opened the pull request:
<!--
  How to contribute: https://vitess.io/docs/contributing/
  Please first make sure there is an open Issue to discuss the feature/fix suggested in this PR.
  If this is a new feature, please mark the Issue as "RFC".
 -->

<!-- if this PR is Work in Progress please create it as a Draft Pull Request -->

## Description
<!-- A few sentences describing the overall goals of the pull request's commits. -->

This is an attempt at making the `vttestserver` container honour a persisted data directory.

There were a few changes required to get here and I'm happy to split into multiple PRs:

* Previously, the `vttestserver` binary did not correctly handle interruption signals. As such, the `defer cluster.TearDown()` was not run in `go/cmd/vttestserver/main.go`. I have added a signal handler to correctly run deferred functions.
* The `mysqlctl` binary was previously always run with the `init` subcommand. We now check if the directory provided in `-data_dir` argument already exists, and if so, we use the `mysqlctl start` command instead!
* I have added a `-keep_data` flag to `vttestserver` binary. This will guard all existing invocations from the new codepath and keeps things backwards compatible.

## Related Issue(s)
<!-- List related issues and pull requests: -->

Fixes https://github.com/vitessio/vitess/issues/7717

## Checklist
- [ ] Should this PR be backported?
    - [x] **I'm not sure!**
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

## Deployment Notes
<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

## Impacted Areas in Vitess
Components that this PR will affect:

- [ ]  Query Serving
- [ ]  VReplication
- [ ]  Cluster Management
- [x]  Build/CI
- [ ]  VTAdmin

deepthi commented on Makefile:
```diff
@@ -251,6 +251,9 @@ docker_local:
 docker_mini:
 	${call build_docker_image,docker/mini/Dockerfile,vitess/mini}
 
+docker_vttestserver:
+	${call build_docker_image,docker/vttestserver/Dockerfile.mysql57,vitess/vttestserver}
+
```
We should build both `mysql57` and `mysql80` images and tag them so. Those are the tags we currently publish on dockerhub too.

deepthi replied:
Sorry if my feedback wasn't clear. As it stands today `make docker_vttestserver` will produce an image like `vitess/vttestserver:latest`. This is confusing because you don't know which mysql flavor you are getting if you just go by the tag. It is in fact the `mysql57` flavor and that fact is embedded in the `Makefile`. To avoid this, we should build `mysql57` and `mysql80` images, and _not_ build a `latest` image at all.
Any change that requires rebuilding one of the images probably requires rebuilding the other. A given user may be using only one of those and not both. If that is what we want to support we should make that explicit.
So we either have two targets `docker_vttestserver_mysql57` and `docker_vttestserver_mysql80`, or just one target `docker_vttestserver` (which builds both images, similar to what you have written as the `docker_vttestserver_all` target).
Another option is to have **three** targets:
```
docker_vttestserver: builds 2 images
docker_vttestserver_mysql57
docker_vttestserver_mysql80
```
That is also acceptable.

hkdsun replied:
oh, I see what you mean 🤦  sorry about that Deepthi. 

I'll address this and attempt to add a test later today.

deepthi reviewed (commented):
This is a nice improvement to `vttestserver`.
We have a unit test and an integration test (`vttest_sample`) for this package.
`vttest_sample` runs in docker on CI. Can you add some tests to cover the new functionality?

hkdsun commented:
I discovered an issue while integrating the new container into our development environment.

The LocalCluster instance actually calls `db.loadSchema()` in its Setup method.

>loadSchema applies sql and vschema migrations respectively for each keyspace in the topology

This becomes problematic across reboots of the test container if the SQL migrations destroy tables – which is the case in the typical output of a Rails application's "schema dump" tool. Something like:

```
DROP TABLE IF EXISTS `test_table`;
CREATE TABLE `test_table` (....) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 ROW_FORMAT=DYNAMIC;
DROP TABLE IF EXISTS `another_table`;
CREATE TABLE `another_table` (....) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 ROW_FORMAT=DYNAMIC;
```

What would the Vitess team want this workflow to look like? I see two options:

1) We short-circuit SQL migrations in the LocalCluster code if running with the `-keep_data` option, or
2) We avoid using destructive migrations and assume that all `.sql` files are run every time the container starts

deepthi commented:
> 
> What would the Vitess team want this workflow to look like? I see two options:
> 
> 1. We short-circuit SQL migrations in the LocalCluster code if running with the `-keep_data` option (except when `Initialize`ing), or
> 2. We avoid using destructive migrations and assume that all `.sql` files are run every time the container starts

Which one is easier for you as a user? I suspect it is the first one.

hkdsun commented:
>Which one is easier for you as a user? I suspect it is the first one.

Correct suspicion indeed 😄 

I have addressed all your feedback, tidied up the code, added the migration short-circuit, and wrote some tests. Would you please have a look at the latest commits?

I found the e2e test in `go/test/endtoend/vtcombo/vttest_sample_test.go` to be redundant compared to the `go/cmd/vttestserver/vttestserver_test.go` but I may be overlooking context. Please let me know if you were expecting coverage in both.

deepthi commented:
I'll re-review, but tests are failing
```
F0323 15:32:34.243946   15971 main.go:252] mkdir /tmp/vttestserver_persistent_mode_510768164/logs: no such file or directory
goroutine 375 [running]:
github.com/golang/glog.stacks(0xc000127500, 0xc0004b8140, 0x7d, 0x96)
	/home/runner/go/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:769 +0xb9
github.com/golang/glog.(*loggingT).output(0x18d1f60, 0xc000000003, 0xc0001291f0, 0x16bdb9e, 0x7, 0xfc, 0x0)
	/home/runner/go/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:720 +0x3b3
github.com/golang/glog.(*loggingT).printDepth(0x18d1f60, 0x3, 0x1, 0xc00012ad90, 0x1, 0x1)
	/home/runner/go/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:646 +0x12d
github.com/golang/glog.(*loggingT).print(...)
	/home/runner/go/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:637
github.com/golang/glog.Fatal(0xc00012ad90, 0x1, 0x1)
	/home/runner/go/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:1128 +0x5c
vitess.io/vitess/go/cmd/vttestserver.runCluster(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/runner/work/vitess/vitess/go/cmd/vttestserver/main.go:252 +0x342
vitess.io/vitess/go/cmd/vttestserver.startCluster(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/runner/work/vitess/vitess/go/cmd/vttestserver/vttestserver_test.go:267 +0x2f8
vitess.io/vitess/go/cmd/vttestserver.TestCanVtGateExecute(0xc000102600)
	/home/runner/work/vitess/vitess/go/cmd/vttestserver/vttestserver_test.go:128 +0xa5
```

deepthi commented:
> I found the e2e test in `go/test/endtoend/vtcombo/vttest_sample_test.go` to be redundant compared to the `go/cmd/vttestserver/vttestserver_test.go` but I may be overlooking context. Please let me know if you were expecting coverage in both.

The endtoend test runs in docker, so it verifies that things work as expected in a docker environment.

hkdsun commented on go/cmd/vttestserver/vttestserver_test.go:
```diff
@@ -243,6 +317,12 @@ func assertEqual(t *testing.T, actual string, expected string, message string) {
-	os.Args = args
+func resetFlags() {
+	os.Args = initArgs
+	config = vttest.Config{}
```
I was a bit baffled about why this passes the tests because I wasn't sure how the `flag` package knows the new pointers to the new struct 😄 

I'm working on it fixing this in a better way but I think it's fairly independent from the rest of the PR.

The reason I was baffled is because of this:


```
package main

import (
 -"flag"
 -"fmt"
 -"os"
)

type Config struct {
 -Dir  string
 -Test string
}

var config Config

func init() {
 -flag.StringVar(&config.Dir, "dir", "/tmp/default", "...")
 -flag.StringVar(&config.Test, "test", "test-default", "...")
}

func main() {
 -initArgs := append([]string{}, os.Args...)
 -os.Args = append(os.Args, "-test=not-default")
 -flag.Parse()
 -fmt.Printf("config = %+v\n", config)

 -os.Args = initArgs
 -os.Args = append(os.Args, "-dir=/tmp/not-default")
 -config = Config{}
 -flag.Parse()
 -fmt.Printf("config = %+v\n", config)
}
```

The program above produces this output:

```
config = {Dir:/tmp/default Test:not-default}
config = {Dir:/tmp/not-default Test:}  // missing its default value of course
```

I still don't understand how the new `config` gets the same pointer as the old `config`.. Must be some memory optimization in go runtime ❓

hkdsun commented:
@deepthi this should be ready for review I think.

The failing test is [flaky on master as well,](https://github.com/vitessio/vitess/actions/runs/679753203) it seems

deepthi commented on Makefile:
```diff
@@ -251,6 +251,13 @@ docker_local:
 
+DOCKER_VTTESTSERVER_SUFFIX = mysql57 mysql80
+DOCKER_VTTESTSERVER_TARGETS = $(addprefix docker_vttestserver_,$(DOCKER_VTTESTSERVER_SUFFIX))
+$(DOCKER_VTTESTSERVER_TARGETS): docker_vttestserver_%:
```
This is a neat trick

deepthi commented on go/cmd/vttestserver/vttestserver_test.go:
```diff
@@ -182,13 +249,21 @@ func TestMtlsAuthUnauthorizedFails(t *testing.T) {
+func startPersistentCluster(dir string, flags ...string) (vttest.LocalCluster, error) {
+	flags = append(flags, []string{
+		"-persistent_mode",
+		// FIXME: if port is not provided, data_dir is not respected
```
The `FIXME` can be removed now?

hkdsun replied:
This is still a bug unfortunately. `-data_dir=` needs to always be passed along with `-port=` or it won't be respected 😬 

It's not really something that concerns us day-to-day but I may give it an hour or two today to make a PR for it anyway, if it's a simple fix

hkdsun replied:
I propose we move forward without this fix in this PR and tackle this work separately. I can open an issue.

deepthi reviewed (approved):
LGTM
